src/AzureClient/AzureClient.cs

```diff
@@ -1,24 +1,249 @@
+            var authority = "https://login.microsoftonline.com/common";
+            var msalApp = PublicClientApplicationBuilder.Create(clientId).WithAuthority(authority).Build();
+
+            // Register the token cache for serialization
+            var cacheFileName = "aad.bin";
+            var cacheDirectoryEnvVarName = "AZURE_QUANTUM_TOKEN_CACHE";
+            var cacheDirectory = System.Environment.GetEnvironmentVariable(cacheDirectoryEnvVarName);
+            if (string.IsNullOrEmpty(cacheDirectory))
+            {
+                cacheDirectory = Path.Join(System.Environment.GetFolderPath(System.Environment.SpecialFolder.UserProfile), ".azure-quantum");
+            }
+
+            var storageCreationProperties = new StorageCreationPropertiesBuilder(cacheFileName, cacheDirectory, clientId).Build();
+            var cacheHelper = await MsalCacheHelper.CreateAsync(storageCreationProperties);
+            cacheHelper.RegisterCache(msalApp.UserTokenCache);
+
+            var scopes = new List<string>() { "https://quantum.microsoft.com/Jobs.ReadWrite" };
+
+            bool promptForLogin = forceLogin;
+            if (!promptForLogin)
+            { 
+                try
+                {
+                    var accounts = await msalApp.GetAccountsAsync();
+                    AuthenticationResult = await msalApp.AcquireTokenSilent(
+                        scopes, accounts.FirstOrDefault()).WithAuthority(msalApp.Authority).ExecuteAsync();
+                }
+                catch (MsalUiRequiredException)
+                {
+                    promptForLogin = true;
+                }
+            }
+
+            if (promptForLogin)
+            {
+                AuthenticationResult = await msalApp.AcquireTokenWithDeviceCode(scopes,
+                    deviceCodeResult =>
+                    {
+                        channel.Stdout(deviceCodeResult.Message);
+                        return Task.FromResult(0);
+                    }).WithAuthority(msalApp.Authority).ExecuteAsync();
+            }
+
+            if (AuthenticationResult == null)
+            {
+                return AzureClientError.AuthenticationFailed.ToExecutionResult();
+            }
+
+            var credentials = new Rest.TokenCredentials(AuthenticationResult.AccessToken);
+
+            QuantumClient = new QuantumClient(credentials);
+            QuantumClient.SubscriptionId = subscriptionId;
+            QuantumClient.ResourceGroupName = resourceGroupName;
+            QuantumClient.WorkspaceName = workspaceName;
+
+            ActiveWorkspace = new Azure.Quantum.Workspace(
+                QuantumClient.SubscriptionId, QuantumClient.ResourceGroupName,
+                QuantumClient.WorkspaceName, AuthenticationResult?.AccessToken);
+
+            try
+            {
+                var jobsList = await QuantumClient.Jobs.ListAsync();
+                channel.Stdout($"Found {jobsList.Count()} jobs in Azure Quantum workspace {workspaceName}");
```

Comment:
```suggestion
                channel.Stdout($"Found {jobsList.Count()} jobs in Azure Quantum workspace {workspaceName}.");
```

Reply:
Yes, good catch, but actually there's no reason to print the count of jobs here. We should just print a message saying successfully connected. (Getting the jobs list was the recommended way to verify that we can connect to the workspace.)

---
In reply to: [427463264](https://github.com/microsoft/iqsharp/pull/144#discussion_r427463264) [](ancestors = 427463264)